### PR TITLE
[feat] DeepNorm/DeepNet support (#227)

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
- Implements [DeepNorm](https://arxiv.org/abs/2203.00555v1) (aka DeepNet paper, 1000 layers Transformer) support, as a "layer norm style" config flag
- cleans up the full model weight init

Open question : should we support something else than xavier init for the weights in the full model case ?

TODO
 - [x] Implement cleanish support
 - [x] Get some curves out on ViT/CIFAR and GPT

 
## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?
- [x] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
cc @jramapuram, I think that at minima the default weight init was probably not the best, but since it was overridden in your example it probably does not explain #219

**Comment by blefaudeux** [comment]
Test with the CIFAR/ViT example, other curves were when testing for #219 : white and light blue are main, deep blue is with this PR, quite the change

gnome-shell-screenshot-5p02s

**Comment by blefaudeux** [comment]
example with the microGPT case: not too much change for 8 layers, but 12 layers and the pre-layer norm case NaNs, while with this PR ("DeepNorm") it's smooth sailing, as claimed by the paper
gnome-shell-screenshot-6kq323

same hyperparams, just changing the layer_norm config

**Comment by blefaudeux** [comment]
> # [Codecov](https://codecov.io/gh/facebookresearch/xformers/pull/230?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) Report
> 
> > Merging [#230](https://codecov.io/gh/facebookresearch/xformers/pull/230?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) ([33a5131](https://github.com/facebookresearch/xformers/commit/33a5131c015984d15b67189cdfa77334af12e9d6)) into [main](https://codecov.io/gh/facebookresearch/xformers/commit/f5c1d014e31e58521977723116cc88736e16d3e9?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) ([f5c1d01](https://github.com/facebookresearch/xformers/commit/f5c1d014e31e58521977723116cc88736e16d3e9)) will **decrease** coverage by `1.11%`.
> > The diff coverage is `41.66%`.
> 
> [Impacted file tree graph](https://codecov.io/gh/facebookresearch/xformers/pull/230?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch)
> 
> ```diff
> @@            Coverage Diff             @@
> ##             main     #230      +/-   ##
> ==========================================
> - Coverage   92.06%   90.94%   -1.12%     
> ==========================================
>   Files          60       60              
>   Lines        3227     3292      +65     
> ==========================================
> + Hits         2971     2994      +23     
> - Misses        256      298      +42     
> ```
> 
> Flag 	Coverage Δ 	
> Python 	`90.94% <41.66%> (-1.12%)` 	arrow_down
> 
> Flags with carried forward coverage won't be shown. [Click here](https://docs.codecov.io/docs/carryforward-flags?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#carryforward-flags-in-the-pull-request-comment) to find out more.
> [Impacted Files](https://codecov.io/gh/facebookresearch/xformers/pull/230?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) 	Coverage Δ 	
> [xformers/factory/model_factory.py](https://codecov.io/gh/facebookresearch/xformers/pull/230/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#diff-eGZvcm1lcnMvZmFjdG9yeS9tb2RlbF9mYWN0b3J5LnB5) 	`78.63% <30.30%> (-19.05%)` 	arrow_down
> [xformers/components/residual.py](https://codecov.io/gh/facebookresearch/xformers/pull/230/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#diff-eGZvcm1lcnMvY29tcG9uZW50cy9yZXNpZHVhbC5weQ==) 	`80.00% <50.00%> (-15.35%)` 	arrow_down
> [xformers/factory/block_factory.py](https://codecov.io/gh/facebookresearch/xformers/pull/230/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#diff-eGZvcm1lcnMvZmFjdG9yeS9ibG9ja19mYWN0b3J5LnB5) 	`89.56% <52.63%> (-4.38%)` 	arrow_down
> 
> [Continue to review full report at Codecov](https://codecov.io/gh/facebookresearch/xformers/pull/230?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/facebookresearch/xformers/pull/230?src=pr&el=footer&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch). Last update [f5c1d01...33a5131](https://codecov.io/gh/facebookresearch/xformers/pull/230?src=pr&el=lastupdated&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch).

extended unit test since that, should be better now

**Comment by blefaudeux** [comment]
ViT/CIFAR example, pre/post/deep norm, same hyper params. The validation accuracy seems likely to end up close (kind of a dummy example), but here again the dynamics seem a fair bit better with deepnorm

Screenshot-20220307224151-990x897

File: examples/microViT.py [hunk 1]
```diff
@@ -248,7 +248,7 @@ def test_step(self, batch, _):
 
     # compute total number of steps
     batch_size = BATCH * GPUS
-    steps = dm.num_samples // batch_size * MAX_EPOCHS
```
Comment (blefaudeux) [inline examples/microViT.py:251]: this should have been fixed earlier, I had a PR stack and somehow lost it with rebases..

File: tests/test_pytorch_transformer_parity.py [hunk 2]
```diff
@@ -172,7 +172,7 @@ def test_pytorch_tranformer_parity(device=torch.device("cuda")):
             dim_feedforward=4 * EMB,
             dropout=DROP,
             activation=ACTIVATION,
-            layer_norm_eps=1e-05,
```
Comment (blefaudeux) [inline tests/test_pytorch_transformer_parity.py:175]: should have been part of https://github.com/facebookresearch/xformers/pull/221 (1e-6 is the default everywhere now), unrelated to this PR but I spotted that while writing this one

File: xformers/components/residual.py [hunk 3]
```diff
@@ -32,20 +34,30 @@ class LayerNormStyle(str, Enum):
 
     Pre = "pre"
     Post = "post"
+    DeepNorm = "deepnorm"
 
 
 # CREDITS: the following is inspired by FastAI's Transformer implementation
 class Residual(nn.Module):
-    """Object-oriented handling of the residual path"""
+    """Object-oriented handling of the residual path.
```
Comment (blefaudeux) [inline xformers/components/residual.py:42]: PR change 1: support scaling the residual path

File: xformers/components/residual.py [hunk 4]
```diff
@@ -86,3 +98,44 @@ def forward(self, inputs: Union[torch.Tensor, List[torch.Tensor]], *args, **kwar
 
         x = self.sublayer(*inputs, *args, **kwargs)
         return self.norm(x)
+
+
+DeepNormCoefficients = namedtuple("DeepNormCoefficients", ["alpha", "beta"])
+
+
+def get_deepnorm_coefficients(
```
Comment (blefaudeux) [inline xformers/components/residual.py:106]: PR change 2: get the residual scaling and init scaling given the whole model, following the paper

File: xformers/factory/block_factory.py [hunk 5]
```diff
@@ -66,27 +67,49 @@ def _get_ln_factory(
...
+        residual: bool,
+        residual_scale: float,
     ):
         if residual:
-            return (
-                Residual(PreNorm(d_model, sublayer, use_triton))
-                if layer_norm_style == LayerNormStyle.Pre
-                else PostNorm(d_model, Residual(sublayer), use_triton)
-            )
+            if layer_norm_style == LayerNormStyle.Pre:
```
Comment (blefaudeux) [inline xformers/factory/block_factory.py:86]: PR change 3: handle 3 layernorm options

File: xformers/factory/model_factory.py [hunk 6]
```diff
@@ -184,22 +187,85 @@ def __init__(
...
+        ), "Reversible layers and deepnorm is not supported for now"
+
+        if use_deepnorm_init:
+            self._deepnorm_weight_init()
+
     @classmethod
     def from_config(cls, config: xFormerConfig):
         return cls(config.stack_configs, config.tie_embedding_weights)
 
+    def _deepnorm_weight_init(self):
```
Comment (blefaudeux) [inline xformers/factory/model_factory.py:203]: PR change 4: handle the required init change, as per the paper

**Comment by blefaudeux** [comment]
cc @stephenroller @suchenzang, just in case you're interested. Small scale tests but seems to confirm the paper indeed

**Comment by blefaudeux** [comment]
rebased on top of https://github.com/facebookresearch/xformers/pull/233, I'll redo some curves for a better comparison

**Comment by blefaudeux** [comment]
> rebased on top of #233, I'll redo some curves for a better comparison

it does not change the accuracy difference really, the NaN case for pre-norm seems gone, but there's still an accuracy difference in that case in favor of deepnorm

**Comment by blefaudeux** [comment]
ping reviewers @fmassa @dianaml0 @jieru-hu
